homeassistant/components/weather/smhi.py

```diff
@@ -0,0 +1,243 @@
+from homeassistant.const import (
+    CONF_LATITUDE, CONF_LONGITUDE,
+    CONF_NAME, TEMP_CELSIUS)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import aiohttp_client
+from homeassistant.util import dt, Throttle
+
+from homeassistant.components.weather import (
+    WeatherEntity, ATTR_FORECAST_CONDITION, ATTR_FORECAST_TEMP,
+    ATTR_FORECAST_TEMP_LOW, ATTR_FORECAST_TIME,
+    ATTR_FORECAST_PRECIPITATION)
+
+from homeassistant.components.smhi.const import (
+    ENTITY_ID_SENSOR_FORMAT, ATTR_SMHI_CLOUDINESS)
+
+DEPENDENCIES = ['smhi']
+REQUIREMENTS = ['smhi-pkg==1.0.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+# Used to map condition from API results
+CONDITION_CLASSES = {
+    'cloudy': [5, 6],
+    'fog': [7],
+    'hail': [],
+    'lightning': [21],
+    'lightning-rainy': [11],
+    'partlycloudy': [3, 4],
+    'pouring': [10, 20],
+    'rainy': [8, 9, 18, 19],
+    'snowy': [15, 16, 17, 25, 26, 27],
+    'snowy-rainy': [12, 13, 14, 22, 23, 24],
+    'sunny': [1, 2],
+    'windy': [],
+    'windy-variant': [],
+    'exceptional': [],
+}
+
+
+# 5 minutes between retrying connect to API again
+RETRY_TIMEOUT = 5*60
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=31)
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Old way of setting up components.
+
+    Can only be called when a user accidentally mentions smhi in the
+    config. In that case it will be ignored.
+    """
+    pass
+
+
+async def async_setup_entry(hass: HomeAssistant,
+                            config_entry: ConfigEntry,
+                            config_entries) -> bool:
+    """Add a weather entity from map location."""
+    location = config_entry.data
+    name = location[CONF_NAME]
+
+    session = aiohttp_client.async_get_clientsession(hass)
+
+    entity = SmhiWeather(name, location[CONF_LATITUDE],
+                         location[CONF_LONGITUDE],
+                         session=session)
+    entity.entity_id = ENTITY_ID_SENSOR_FORMAT.format(name)
+
+    config_entries([entity], True)
+    return True
+
+
+class SmhiWeather(WeatherEntity):
+    """Representation of a weather entity."""
+
+    def __init__(self, name: str, latitude: str,
+                 longitude: str,
+                 session: aiohttp.ClientSession = None) -> None:
+        """Initialize the SMHI weather entity."""
+        from smhi import Smhi
+
+        self._name = name
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecasts = None
+        self._fail_count = 0
+        self._smhi_api = Smhi(self._longitude, self._latitude,
+                              session=session)
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    async def async_update(self) -> None:
+        """Refresh the forecast data from SMHI weather API."""
+        from smhi.smhi_lib import SmhiForecastException
+
+        def fail():
+            self._fail_count += 1
```

Comment:
If it is marked as unavailable there shouldn't be any bad values

Reply:
Thank you for your review @dersger. The forecast[0] will have the current as last requested from API. The lib returns a consolidated day by day forecast so the next one will be the next day forecast basically. Marking it as unavailable is a option for sure but I am not totally convinced it would be a great idea for a failed update for weather cause I rather have one or two hours of slightly wrong weather data than no data att all?   It seems reasonable both ways. I could live with the function as for now. 
A future enhancement would be to rewrite the lib to return like 5 hours of additional weather data, at update find the one that are valid (usually the first one) or one of the following 4 if the update went wrong and after all 5 hours of forecasts are not in valid time the weather is considered "unavailable".

Reply:
Will consider this in future PR. Thanks for the suggestion.